src/ucs/datastruct/arbiter.c

```diff
@@ -329,7 +333,7 @@ void ucs_arbiter_dispatch_nonempty(ucs_arbiter_t *arbiter, unsigned per_group,
                 elem->list.next = elem_list_next;
                 /* make sure that next dispatch() will continue
                  * from the current group */
-                arbiter->current = group_head;
+                arbiter->current = elem;
```

Comment:
can we add assert that elem==group_head?

Reply:
```elem==group_head``` this is incorrect assumption, there is valid flow when group_head doesn't point to real group head

Reply:
i see
so there is another issue: group_head is used in several places, after it could already be removed. probably not causing a bug because it's just for comparing with prev_group
nevertheless, maybe let's also set `is_single_group = group_head==prev_group` before the loop, to avoid using group_head at all inside the loop?

Reply:
done